sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,264 @@
+
+                return IsAllowedEnumerableValue(types[1], ((IDictionary)value).Values);
+            }
+
+            private static bool IsAllowedEnumerableValue(Type elementType, IEnumerable enumerable)
+            {
+                foreach (var item in enumerable)
+                {
+                    if (item == null)
+                    {
+                        continue;
+                    }
+
+                    // Don't allow heterogenous collections
+                    if (item.GetType() != elementType)
+                    {
+                        return false;
+                    }
+
+                    if (!IsAllowedValue(item))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private static bool IsAllowedAnonymousValue<T>(Type type, T value, Type[]? visited)
+            {
+                if (!IsAnonymousType(type))
+                {
+                    return false;
+                }
+
+                foreach (PropertyInfo property in type.GetProperties())
+                {
+                    if (IsAllowedLeafType(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    object? propertyValue = property.GetValue(value);
+                    if (IsAllowedCollectionValue(property.PropertyType, propertyValue))
+                    {
+                        continue;
+                    }
+
+                    // Detect cycles: trust but verify
+                    if (ContainsType(visited, property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    // Recurse
+                    visited = AddType(visited, type);
+                    if (!IsAllowedAnonymousValue(property.PropertyType, propertyValue, visited))
+                    {
+                        return false;
+                    }
+
+                    ArrayPool<Type>.Shared.Return(visited);
+                }
+
+                return true;
+            }
+
+            private static bool IsAnonymousType(Type type)
+            {
+                return type.Name.StartsWith("<>f__AnonymousType");
+            }
+
+            private static bool ContainsType(Type[]? types, Type type)
+            {
+                if (types == null)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < types.Length; i++)
+                {
+                    if (types[i] == type)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private static Type[] AddType(Type[]? types, Type type)
+            {
+                int length = types == null ? 0 : types.Length;
+                Type[] expanded = ArrayPool<Type>.Shared.Rent(length + 1);
```

Comment:
We cannot rent every time a type is added

Reply:
Ok, I'll allocate something larger to start.

Reply:
Addressed.